client/blb/client_test.go

```diff
@@ -257,6 +304,376 @@ func TestWriteReadTracedExactlyOneTract(t *testing.T) {
+		testRead(t, blob, core.TractLength, core.TractLength, readDone)
+		readDone <- true
+	}()
+
+	// let 2 go first
+	delayChans[1] <- time.Time{}
+	<-rdone
+	trace.checkRead(t, "ts-0000000100000001:0001-1", 1, core.TractLength, 0)
+
+	// check if we cancel the others and the read completes.
+	<-cancelCh
+	<-bdone
+	<-readDone
+
+	trace.checkReadAbsence(t, "ts-0000000100000001:0001-0", 1, core.TractLength, 0)
+	trace.checkReadAbsence(t, "ts-0000000100000001:0001-2", 1, core.TractLength, 0)
+}
+
+func TestBackupReadThirdFinishes(t *testing.T) {
+	cli, trace := newTracingClient()
+	defer cli.restoreTestFuncs()
+
+	// Write a blob and check the writes are logged.
+	blob := createBlob(t, cli)
+	testWrite(t, blob, core.TractLength, core.TractLength)
+
+	trace.check(t, true, "ts-0000000100000001:0000-0", 1, 0, 0)
+	trace.check(t, true, "ts-0000000100000001:0000-1", 1, 0, 0)
+	trace.check(t, true, "ts-0000000100000001:0000-2", 1, 0, 0)
+	trace.check(t, true, "ts-0000000100000001:0001-0", 1, core.TractLength, 0)
+	trace.check(t, true, "ts-0000000100000001:0001-1", 1, core.TractLength, 0)
+	trace.check(t, true, "ts-0000000100000001:0001-2", 1, core.TractLength, 0)
+
+	// Test one request per host, first request finishes
+	rdone, bdone, cancelCh := cli.setupBackupClient(3, true) // 2 backup requests
+	delayChans := setupBackupRequestFunc(cli, 3)
+	readDone := make(chan bool)
+	go func() {
+		blob.Seek(core.TractLength, os.SEEK_SET)
+		testRead(t, blob, core.TractLength, core.TractLength, readDone)
+		readDone <- true
+	}()
+
+	// relase the thrid reader first
+	delayChans[2] <- time.Time{}
+	<-rdone
+
+	trace.checkRead(t, "ts-0000000100000001:0001-2", 1, core.TractLength, 0)
+
+	<-cancelCh
+	<-bdone
+	<-readDone
+
+	trace.checkReadAbsence(t, "ts-0000000100000001:0001-0", 1, core.TractLength, 0)
+	trace.checkReadAbsence(t, "ts-0000000100000001:0001-1", 1, core.TractLength, 0)
+}
+
+func TestReadFailoverErrOnFirst(t *testing.T) {
+	fail := func(e tsTraceEntry) core.Error {
+		// Reads from the first tractserver fails.
+		if !e.write && strings.HasSuffix(e.addr, "0") {
+			return core.ErrRPC
+		}
+		return core.NoError
+	}
+	cli := newClient(fail)
+	defer cli.restoreTestFuncs()
+
+	// Write a blob and check the writes are logged.
+	blob := createBlob(t, cli)
+	testWrite(t, blob, core.TractLength, core.TractLength)
+
+	// Test one request per host, first request finishes
+	rdone, bdone, cancelCh := cli.setupBackupClient(3, true)
+	delayChans := setupBackupRequestFunc(cli, 3)
+	readDone := make(chan bool)
+	go func() {
+		blob.Seek(core.TractLength, os.SEEK_SET)
+		testRead(t, blob, core.TractLength, core.TractLength, readDone)
+		readDone <- true
+	}()
+
+	// relase the first request first
+	delayChans[0] <- time.Time{}
+	<-rdone
+
+	// since the first one returns an error, try the next.
+	delayChans[1] <- time.Time{}
+	<-rdone
+
+	// this one passes, so it will cancel the last request.
+	<-cancelCh
+	<-bdone
+	<-readDone
+}
```

Comment:
should this (and all the following tests) check the trace log too? how do they tell that the right thing is happening?